src/ppx/scene/scene_node.cpp

```diff
@@ -0,0 +1,206 @@
+#include "ppx/scene/scene_node.h"
+
+namespace ppx {
+namespace scene {
+
+// -------------------------------------------------------------------------------------------------
+// Node
+// -------------------------------------------------------------------------------------------------
+Node::Node(scene::Scene* pScene)
+    : mScene(pScene)
+{
+}
+
+Node::~Node()
+{
+}
+
+void Node::SetVisible(bool visible)
+{
+    mVisible = visible;
+}
+
+const float4x4& Node::GetEvaluatedMatrix() const
+{
+    if (mEvaluatedDirty) {
+        float4x4 parentEvaluatedMatrix = float4x4(1);
+        if (!IsNull(mParent)) {
+            parentEvaluatedMatrix = mParent->GetEvaluatedMatrix();
+        }
+        auto& concatenatedMatrix = GetConcatenatedMatrix();
+        mEvaluatedMatrix         = parentEvaluatedMatrix * concatenatedMatrix;
+        mEvaluatedDirty          = false;
+    }
+    return mEvaluatedMatrix;
+}
+
+void Node::SetEvaluatedDirty()
+{
+    mEvaluatedDirty = true;
+    for (auto& pChild : mChildren) {
+        pChild->SetEvaluatedDirty();
+    }
+}
+
+void Node::SetTranslation(const float3& translation)
+{
+    Transform::SetTranslation(translation);
+    SetEvaluatedDirty();
+}
+
+void Node::SetRotation(const float3& rotation)
+{
+    Transform::SetRotation(rotation);
+    SetEvaluatedDirty();
+}
+
+void Node::SetScale(const float3& scale)
+{
+    Transform::SetScale(scale);
+    SetEvaluatedDirty();
+}
+
+void Node::SetRotationOrder(Transform::RotationOrder value)
+{
+    Transform::SetRotationOrder(value);
+    SetEvaluatedDirty();
+}
+
+void Node::SetParent(scene::Node* pNewParent)
```

Comment:
I looked at the logic of having an either/or case of SetParent/AddChild and it got a bit too much to handle. I simplified it so that the calling code has more responsibility of of parenting:
 - A node can only add a child node if that child node doesn't have an existing parent and node isn't in the child's subtree
 - Calling code is responsible for asking parent to remove child if it wants to reparent